### ISSUE TITLE
Fix Windows support in twisted.python.lockfile for Python 3.x

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6628]
|<img alt="hbock's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| hbock reported|
|-|-|
|Trac ID|trac#6628|
|Type|defect|
|Created|2013-07-15 01:29:47Z|
|Branch|https://github.com/twisted/twisted/tree/lockfile-windows-py3-compatibility-6628-2|

twisted.python.lockfile is broken on Python 3.x/Windows because of several issues:

 - Similar to [#6627](https://github.com/twisted/twisted/issues/6627), the exception handling of IOError/OSError/WindowsError causes issues on Python 3.x, and it is no longer necessary on Python 2.6+.
 - **file** is no longer a builtin in Python 3.x, and some windows-specific code relies on it.  Also, using **file** as a way to open files is no longer recommended even in 2.x code.
 - the open() builtin in Python 3.x does not support the commit ('c') mode that is specific to the MSVCRT.  Unfortunately there doesn't seem to be a way around it without calling into the CRT via ctypes or an extension, or possibly using win32api instead, if we want to retain atomicity.  Further investigation is required.

Attachments:

* [twisted.python.filepath-6628-1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/c56/c5646c24aae34705a73634c70f2616d6428d2a77/1d2f4ebdac33740811f930f777207a0918075d42.patch) (2603 bytes) - added by hbock on 2013-07-15 01:31:43Z - suggested patch, attempt #1

<details><summary>Searchable metadata</summary>

```
trac-id__6628 6628
type__defect defect
reporter__hbock hbock
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_lockfile_windows_py3_compatibility_6628_2 branches/lockfile-windows-py3-compatibility-6628-2
branch_author__hawkowl__rwall hawkowl, rwall
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1373851787000000 1373851787000000
changetime__1442980486923766 1442980486923766
version__None None
owner__hawkowl hawkowl
cc__hbock
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="hbock's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|hbock commented|
|-|-|

This patch resolves the issues on Windows, allowing unittests to run clean on both 2.6 and 3.3.

It does not address the issue of not being able to guarantee that flush() immediately writes out to disk on 3.3.  After some initial research, it appears using win32file.CreateFile with dwFlagsAndAttributes set to FILE_FLAG_WRITE_THROUGH and calling FlushFileBuffers() might provide similar guarantees as fopen(xxx, "wcb").

I'm curious as to what the 'c' really affords us in making the symlink operation atomic - wouldn't the rename() operation be the one that's important? I would think FILE_FLAG_WRITE_THROUGH would only help if someone other process was waiting on the file write via FindFirstChangeNotification or similar, but that is not the case here.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_3"></a>|@wallrj commented|
|-|-|

(In [39961]) Branching to 'lockfile-windows-py3-compatibility-6628'

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_4"></a>|@wallrj commented|
|-|-|

(In [39962]) Apply twisted.python.filepath-6628-1.patch from hbock. refs [#6628](https://github.com/twisted/twisted/issues/6628)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_5"></a>|@wallrj commented|
|-|-|

Thanks hbock,

I applied your patch to a branch and forced a build. 

The results look good (apart from some spurious pyflakes warnings):
 * https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/lockfile-windows-py3-compatibility-6628

I'm not a windows user so can't comment on the code or answer your comments (above).

A general comment is that your patch doesn't have any new or changed tests.

And yet you say that it "allows unittests to run clean on both 2.6 and 3.3."

So are there some tests that are currently skipped on windows that now need enabling?

Are there some tests that are not currently included in the python3 test suite that need adding?

I don't think we run the python3 tests on any of our Windows builders. That probably needs addressing. I created a ticket:
 * https://github.com/twisted-infra/buildslave-config/issues/1

Please comment on that if you have any thoughts about running python3 tests on Windows. 

I also followed the MSDN link about the "commit" fopen flag -- but it seems to suggest that the flag is available. Is there a link that explains clearly why its not available in python3? I found this one:
 * http://bugs.python.org/issue12103#msg136225


So anyway, I'll leave this in review and hopefully someone who knows windows can do a proper review.

Meanwhile if you want to update your patch please make it against the new branch.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

It sounds like this is blocked on a Python 3 buildbot for Windows.  It doesn't look like there's much windows-specific in this patch other than dropping the "commit" flag from the file open, which looks like we might be losing some atomicity guarantees?  Perhaps someone needs to open a bug on CPython.

Thanks for your contribution, hbock, I hope someone steps forward with some builder resources soon.  In the meanwhile, we need to do some work to make our existing py3 builder turn green.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph commented|
|-|-|

I can't seem to reassign this to hbock, trac doesn't think they're a real user :(. I hope they'll get an email about it and click "accept"…

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_8"></a>|@hawkowl commented|
|-|-|

(In [45691]) Branching to lockfile-windows-py3-compatibility-6628-2.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_9"></a>|@hawkowl commented|
|-|-|

So, testing on Windows + Py3 is blocked on this issue.

I looked into the 'atomic' thing and even read some Py3 source code (ew) and win32api -- nothing applicable seems to be in there, and Py3 has changed to make it effectively impossible (using flags, and there is no flag for it, only a character mode).

I've polished up this branch a little. Builders spun, please review.

The results of the 3.5 builder is here: https://buildbot.twistedmatrix.com/builders/windows7-64-py3.5/builds/10/steps/select/logs/stdio The error does not exist now, and we have some brand new exciting errors!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_10"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Thanks for working on this.

Can we make a public statement that lockfile works with Py3 on Windows?
Do we have automated tests for that? Do those tests pass? :)

From what I read in this ticket, this is a blocker to have a py3 buildslave running on Windows.

I think that is fine to merge it, but just make it a .misc ticket, and create a new ticket to check that all test pass.
Once the test pass we can update the release notes.

----------

Since we are touching this code, I think that is a good idea to document the fake symlink/readlink method. From the current method name it is not clear that they don't implement read symlinks (ex reparse points). From what I can see they are just a hack used by the locking system.

----------

Please check my comments and merge.

Looking forward for having the py3 slave up and running on Windows :)

Many thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_11"></a>|@hawkowl set status to `closed`|
|-|-|

(In [45700]) Merge lockfile-windows-py3-compatibility-6628-2: Make t.p.lockfile work on Windows

Authors: hbock, hawkowl
Reviewers: rwall, glyph, adiroiban
Fixes: [#6628](https://github.com/twisted/twisted/issues/6628)
